### ISSUE TITLE
'generator' object has no attribute 'tag'

### ISSUE THREAD

**Issue by martin-xia** [ecederstrand/exchangelib#286]
Hi there,

This is a great library, easy to use, pretty stable. But i've got this issue recently:
```python
File "/usr/local/lib/python2.7/dist-packages/exchangelib/account.py", line 66, in __init__
    credentials=credentials)
  File "/usr/local/lib/python2.7/dist-packages/exchangelib/autodiscover.py", line 182, in discover
    protocol=protocol)
  File "/usr/local/lib/python2.7/dist-packages/exchangelib/autodiscover.py", line 312, in _autodiscover_quick
    r = _get_autodiscover_response(protocol=protocol, email=email)
  File "/usr/local/lib/python2.7/dist-packages/exchangelib/autodiscover.py", line 347, in _get_autodiscover_response
    data = _get_autodiscover_payload(email=email)
  File "/usr/local/lib/python2.7/dist-packages/exchangelib/autodiscover.py", line 343, in _get_autodiscover_payload
    return xml_to_str(payload, encoding=DEFAULT_ENCODING, xml_declaration=True)
  File "/usr/local/lib/python2.7/dist-packages/exchangelib/util.py", line 117, in xml_to_str
    ElementTree(tree).write(stream, encoding=encoding, xml_declaration=False, method='xml')
  File "/usr/lib/python2.7/xml/etree/ElementTree.py", line 820, in write
    serialize(write, self._root, encoding, qnames, namespaces)
  File "/usr/lib/python2.7/xml/etree/ElementTree.py", line 939, in _serialize_xml
    _serialize_xml(write, e, encoding, qnames, None)
  File "/usr/lib/python2.7/xml/etree/ElementTree.py", line 900, in _serialize_xml
    tag = elem.tag
AttributeError: 'generator' object has no attribute 'tag'
```

Do u you have any suggestion or idea why is it failing? 

Thanks,

**Comment by martin-xia** [comment]
Also when i turn off auto-discover, it shows with the same exception:

File "/usr/local/lib/python2.7/dist-packages/exchangelib/account.py", line 83, in __init__
    self.root = Root.get_distinguished(account=self)
  File "/usr/local/lib/python2.7/dist-packages/exchangelib/folders.py", line 537, in get_distinguished
    folders = list(cls.get_folders(account=account, ids=[cls(account=account, name=cls.DISTINGUISHED_FOLDER_ID)]))
  File "/usr/local/lib/python2.7/dist-packages/exchangelib/folders.py", line 528, in get_folders
    shape=IdOnly
  File "/usr/local/lib/python2.7/dist-packages/exchangelib/services.py", line 876, in call
    shape=shape,
  File "/usr/local/lib/python2.7/dist-packages/exchangelib/services.py", line 76, in _get_elements
    response = self._get_response_xml(payload=payload)
  File "/usr/local/lib/python2.7/dist-packages/exchangelib/services.py", line 133, in _get_response_xml
    soap_payload = wrap(content=payload, version=api_version, account=account)
  File "/usr/local/lib/python2.7/dist-packages/exchangelib/transport.py", line 67, in wrap
    return xml_to_str(envelope, encoding=DEFAULT_ENCODING, xml_declaration=True)
  File "/usr/local/lib/python2.7/dist-packages/exchangelib/util.py", line 117, in xml_to_str
    ElementTree(tree).write(stream, encoding=encoding, xml_declaration=False, method='xml')
  File "/usr/lib/python2.7/xml/etree/ElementTree.py", line 820, in write
    serialize(write, self._root, encoding, qnames, namespaces)
  File "/usr/lib/python2.7/xml/etree/ElementTree.py", line 939, in _serialize_xml
    _serialize_xml(write, e, encoding, qnames, None)
  File "/usr/lib/python2.7/xml/etree/ElementTree.py", line 900, in _serialize_xml
    tag = elem.tag
AttributeError: 'generator' object has no attribute 'tag'

**Comment by ecederstrand** [comment]
Hmm, I have no obvious solutions to this. Can you post some example code that causes this? Also, which version of `lxml` are you using?

**Comment by martin-xia** [comment]
Here's the code:

                   ```
               if not use_autodiscover:
                        config = Configuration(server=server_name, credentials=credentials)
                        account = Account(
                            primary_smtp_address=username,
                            config=config,
                            autodiscover=False,
                            access_type=DELEGATE
                        )
              else:
                        account = Account(
                            primary_smtp_address=username,
                            credentials=credentials,
                            autodiscover=True,
                            access_type=DELEGATE
                        )
            
            m = Message(
                account=account,
                subject=subject,
                body=HTMLBody(msgHtml),
                to_recipients=[Mailbox(email_address=to)]
            )
            m.send_and_save()

```

And here's the other thing: lxml==4.0.0

Hope it helps, thx!

**Comment by ecederstrand** [comment]
Hmm, that's quite straight-forward code, and tests are passing on TravisCI with lxml==4.0.0.

Which version of exchangelib are you using? Does this also happen on 1.10.4?

**Comment by martin-xia** [comment]
I'm currently using exchangelib==1.10.3
However, i did see the same problem when i'm using older versions...

Actually i'm running those code via celery's async task, does it make any difference? And by the way the problem i have is not always there, it just happens from time to time, which makes it even worse.  I have to restart my celery everytime i get the problem.

Thanks

**Comment by ecederstrand** [comment]
What happened is that somewhere in the XML tree we generate, a generator sneaked in that should have been an `Element` type. I have no idea how that happened, and I cannot reproduce it, not even when ignoring exchangelib and using just stuff from the `xml.etree` module. It may be caused by a bug in exchangelib, or it may be due to some bad input value on your side.

I think your best bet is to instrument /usr/lib/python2.7/xml/etree/ElementTree.py line 900 with something like:

```python
def _serialize_xml(write, elem, encoding, qnames, namespaces):
    try:
        tag = elem.tag
    except AttributeError:
        raise AttributeError('Expected Element, got generator: %s' % list(elem))
```

Maybe the value of `elem` will give you a hint about what's happening.

**Comment by martin-xia** [comment]
Yes, i've tried to do that in the exact same line. The problem is it's almost impossible to reproduce this problem. It happened twice in the past 6 month, which actually make the problem even worse...

Now i'm thinking to catch the error and restart the celery server...

**Comment by martin-xia** [comment]
Thanks anyway, plz let me know if you find anything else

**Comment by martin-xia** [comment]
Hmm, actually i'm thinking to do this:

```
try:
     do my own stuff blah
except AttributeError as e:
     util._deepcopy_cache.clear()
```
because i'm suspecting the cache is compromised.

Do you see any problem other than efficiency?

Thanks

**Comment by ecederstrand** [comment]
Ouch, a bug that takes 3 months to reproduce. I feel you pain :-)

You last suggestion looks OK, but since you mentioned celery, you should probably acquire a `multiprocessing.Lock` before clearing, to avoid race conditions between threads.

Looking at the `_deepcopy_cache` usage, I see a bug in `create_element()` We're checking for existence of key `name` but then iserting key `key`. I can't think of how it could trigger your bug, but it's definitely wrong. Running tests on a patch now.

**Comment by ecederstrand** [comment]
The referenced commit added an `exchangelib.util._deepcopy_cache_lock`, so your code could do the following, which is safer:

```python
try:
     do my own stuff blah
except AttributeError as e:
    with util._deepcopy_cache_lock:
        util._deepcopy_cache.clear()
```

**Comment by martin-xia** [comment]
Thx for this PR, so do you know when can i get the new version with this fix? 

Although i agree with you that this is not the cause of my problem, but i'm desperate to try anything now :-D

**Comment by ecederstrand** [comment]
I hadn't planned a new release just yet, but you can easily install this exact commit from GitHub until then:

```bash
pip install git+https://github.com/ecederstrand/exchangelib.git@cd891106ed50eaaf72c6e5307802c95c95bbd048
```

**Comment by ecederstrand** [comment]
Just released v1.10.6. Feel free to re-open if you're able to reproduce this issue.
